### ISSUE TITLE
Tab/Space issues?

### ISSUE THREAD

**Issue by gotbletu** [junegunn/fzf#118]
not sure how to describe this but seems like fzf has issues to tabs/spaces inside text files or it does something to change the output?

how do you fix that? I try on other programs like "percol" and it came out fine as you can see from the picture.

See the **green line**
http://i.imgur.com/U1Y9BGO.jpg

    #fzf with surfraw
    surfraw -browser=$BROWSER "$(cat ~/.config/surfraw/bookmarks | sed '/^$/d' | sort -n | fzf -e)"
    #percol with surfraw
    surfraw -browser=$BROWSER "$(cat ~/.config/surfraw/bookmarks | sed '/^$/d' | sort -n | percol)"

**Comment by junegunn** [comment]
fzf does not do any processing on tab characters, and ncurses sets a tab stop every 8 columns by default. I once considered making it configurable, but discarded the idea because of compatibility issues (It didn't run on old linux distros; https://github.com/junegunn/fzf/commit/3ed86445)

```sh
# I use 8-col tabs on terminal, so it comes out like this
ruby -e '(1..20).each { |e| puts "-" * e + "\t1" }'

-       1
--      1
---     1
----    1
-----   1
------  1
------- 1
--------        1
---------       1
----------      1
-----------     1
------------    1
-------------   1
--------------  1
--------------- 1
----------------        1
-----------------       1
------------------      1
-------------------     1
--------------------    1

# ncurses, which fzf depends on, also uses 8-col tabs, but because of the
# 2-col indentation of fzf, the output looks slightly different
ruby -e '(1..20).each { |e| puts "-" * e + "\t1" }' | fzf

  --------------------  1
  -------------------   1
  ------------------    1
  -----------------     1
  ----------------      1
  ---------------       1
  --------------        1
  ------------- 1        
  ------------  1        
  -----------   1        
  ----------    1        
  ---------     1        
  --------      1        
  -------       1        
  ------        1        
  ----- 1                
  ----  1                
  ---   1                
  --    1                
> -     1                
  20/20                  
>                        
```

I've just downloaded peco and it looks like it enforces 4-character tab stops.

```sh
ruby -e '(1..20).each { |e| puts "-" * e + "\t1" }' | peco

QUERY>                                                                                                                                                       IgnoreCase [20 (1/1)]
-   1
--  1
--- 1
----    1
-----   1
------  1
------- 1
--------    1
---------   1
----------  1
----------- 1
------------    1
-------------   1
--------------  1
--------------- 1
----------------    1
-----------------   1
------------------  1
------------------- 1
--------------------    1
```

Because of the aforementioned compatibility issue, I probably won't make tab width configurable. Maybe you could try to use a preprocessor which aligns the columns, like follows?

https://gist.github.com/junegunn/b3386e912ee9229824ef

```sh
ruby -e '(1..20).each { |e| puts "-" * e + "\t1" }' | ./align.rb
-                    1
--                   1
---                  1
----                 1
-----                1
------               1
-------              1
--------             1
---------            1
----------           1
-----------          1
------------         1
-------------        1
--------------       1
---------------      1
----------------     1
-----------------    1
------------------   1
-------------------  1
-------------------- 1
```

**Comment by junegunn** [comment]
Is the program in the first image Vim? On Vim, you can change tab stops using `:set ts=8`, try some different values and you'll get a better idea of what's going on.

**Comment by junegunn** [comment]
peco manually sets tab stops by 4 columns:
https://github.com/peco/peco/blob/74e0e70/layout.go#L69-L75

For future reference (for me), conditional compilation based on the version of ncurses:
https://gitorious.org/tig/mainline/commit/3ff7fff834173c6d01f04feef6403c6b0a724ebf

**Comment by gotbletu** [comment]
I tried different tabstop numbers in vim, doesnt seem to do anything to fix it. Still output not aligned like the original text file.

have not used peco, but i was using percol
https://github.com/mooz/percol

**Comment by junegunn** [comment]
Ah sorry my bad, peco is a Go-implementation of percol. Percol also seems to 8-col tab size like fzf

```sh
ruby -e '(1..20).each { |e| puts "-" * e + "\t1" }' | percol

QUERY>                                                                                                                                                                (1/20) [1/1]
-       1
--      1
---     1
----    1
-----   1
------  1
------- 1
--------        1
---------       1
----------      1
-----------     1
------------    1
-------------   1
--------------  1
--------------- 1
----------------        1
-----------------       1
------------------      1
-------------------     1
--------------------    1
```

Could you post the text file on gist or somewhere?

**Comment by gotbletu** [comment]
text for my bookmarks

http://slexy.org/raw/s20mjCmaRz

**Comment by junegunn** [comment]
Thanks, now I see the reason. It's because of 2-column indentation of fzf interferes with tab stops (note that percol does not have left-side indentation).

The default fzf with 2-space indentation:

![with-indent](https://cloud.githubusercontent.com/assets/700826/5791403/89fdea96-9f17-11e4-85a1-e4ebd8665f93.png)

Modified version of fzf w/o indentation:

![wo-indent](https://cloud.githubusercontent.com/assets/700826/5791405/a15083d4-9f17-11e4-95da-0425a3c83a2f.png)

Using the aforementioned [align script](https://gist.github.com/junegunn/b3386e912ee9229824ef):

```
cat input | align.rb | fzf
```

![aligned](https://cloud.githubusercontent.com/assets/700826/5791414/06976bb8-9f18-11e4-82c0-5a5e2c42591c.png)

Maybe I could add an option to remove 2-space indentation on the left. What do you think?

**Comment by gotbletu** [comment]
seems like the "Modified version of fzf w/o indentation" looks best. Yea maybe just add extra flag to fix it; lols fzf -notabdeath -e

**Comment by junegunn** [comment]
The left gutter is important when using `--multi` option where you can select multiple items. If I'd want to remove the gutter, I have to come up with a way to indicate whether an item is selected or not. I'll think about it.

**Comment by junegunn** [comment]
Instead of adding a new option or finding a way to customize ncurses, I decided to manually process tab characters. Here's what it looks like:

![0 9 2-dev](https://cloud.githubusercontent.com/assets/700826/5791791/f07f86fa-9f33-11e4-9f29-f078ebde65b3.png)

I'll be releasing 0.9.2 in a week or two, until then, you can download 0.9.2-dev binary with the change from the following link:
https://github.com/junegunn/fzf-bin/releases/tag/0.9.2-dev

**Comment by gotbletu** [comment]
Thanks, works find & looks pretty now =)

**Comment by junegunn** [comment]
I've just released 0.9.2 with a few more fixes:
https://github.com/junegunn/fzf-bin/releases/tag/0.9.2
